### ISSUE TITLE
Enhanced Timer Usability

### ISSUE THREAD

**Issue by rallen2010** [flowsforapex/apex-flowsforapex#248]
This feature collects several issues into a meaningful feature.
- Most APEX developers not familiar / comfortable with ISO time strings
- Cycle timers not working - should work for non-interrupting timer boundary events (#226 , #184 )
- No Linting to ensure timer definitions have been set for timer events

1. Add 'Date (Oracle format)' timer definition type that allows you to specify an Oracle date string and an Oracle format mask.  
2. Add 'Duration (Oracle format)' timer definition type that allows you to specify an Oracle Time Interval (Days to Seconds) using an appropriate interval format mask. 
3. Add 'Cycle (Oracle format)' cycle timer that allows you to specify a Duration (oracle format)  for first occurrence, an oracle-format duration for interval time, and a number of cycles, along with the  Oracle format masks to be used. 
4. Fix cycle timers for Non Interrupting Timer Boundary Events only (allows repeated reminders).  On a Cycle Timer, the 'before-event' variable expression will fire once, before the first timer is set; the 'on-event' variable expression set will fire  when each repeated timer fires.
5. Add Linting to ensure timer events have a timer definition specified
6. Current timer implementation has no error checking or error messages if the timer definition is not well formed.
7. For bonus marks, it would be nice to be able to change the run time of a waiting task to a different time. (business needs change sometimes, so things need to move forwards or backwards...)

**Comment by rallen2010** [comment]
XML sand flow_object_attribute table formats / specs now detailed in https://github.com/mt-ag/apex-flowsforapex/wiki/Timer-Definitions-and-their-Keys-in-flow_object_attributes

**Comment by kannankumara** [comment]
start_timer procedure in flow_timer_pkg has a typo error in the following, please change 6 to 5 as shown below:
OLD: substr(l_timer_def,6,length(l_timer_def)-**6**)
NEW: substr(l_timer_def,6,length(l_timer_def)-**5**)

**Comment by rallen2010** [comment]
@kannankumara 
Thanks for your comment.  What version are you using?  V21.1?
Can I just check whether your variable that you are trying to substitute has been specified using the correct syntax?  

We expect variables to be included like this `&F4A$myDuration.` (please check that you have the ` .` at the end).
This is in line with so called exact substitutions which are now standard elsewhere in APEX.
This was the reason for cutting out the last character in above code.   (see previous issue #200).

I would propose you adapt your definition accordingly as the same will be required with the revised variable substitution system.

If not, please let me know.

> start_timer procedure in flow_timer_pkg has a typo error in the following, please change 6 to 5 as shown below:
> OLD: substr(l_timer_def,6,length(l_timer_def)-**6**)
> NEW: substr(l_timer_def,6,length(l_timer_def)-**5**)

**Comment by kannankumara** [comment]
Richard, thank you for clarifying and providing the link to #200. It makes sense as I had indeed missed the last "." as I didn't relate that format to the standard APEX substitution syntax. I was wondering why you guys introduced the new format of starting the variable with & and a $ in between!

Now it makes perfect sense, so I'll revert the change I made to the package and use the standard APEX substitution string syntax (of using the "." at the end".
